rpc: Add a parameter to sendrawtransaction which sets a maximum value for unspendable outputs.

davidgumberg opened the pull request:
This PR adds a user configurable, zero by default parameter — `maxburnamount` — to `sendrawtransaction`. This PR makes bitcoin core reject transactions that contain unspendable outputs which exceed `maxburnamount`.  closes #25899.

As a result of this PR, `sendrawtransaction` will by default block 3 kinds of transactions:

1. Those that begin with `OP_RETURN` - (datacarriers)
2. Those whose lengths exceed the script limit.
3. Those that contain invalid opcodes.

The user is able to configure a `maxburnamount` that will override this check and allow a user to send a potentially unspendable output into the mempool.

I see two legitimate use cases for this override:
1. Users that deliberately use `OP_RETURN` for datacarrier transactions that embed data into the blockchain.
2.  Users that refuse to update, or are unable to update their bitcoin core client would be able to make use of new opcodes that their client doesn't know about.

1440000bytes commented:
Concept ACK

I am not sure about name of this argument: `rejectreturn`

kristapsk commented:
Concept ACK

brunoerg commented on src/rpc/mempool.cpp:
```diff
@@ -61,12 +62,24 @@ static RPCHelpMan sendrawtransaction()
+                UniValue::VBOOL
             });
 
+            bool reject_return = request.params[2].isNull() ? true : request.params[2].isTrue();
```
```suggestion
            const bool reject_return = request.params[2].isNull() ? true : request.params[2].isTrue();
```

davidgumberg replied:
I have pushed this

brunoerg commented:
Concept ACK

I agree on @1440000bytes about the name of the argument `rejectreturn`, I think this name is not so intuitive about what it does.

davidgumberg commented:
@1440000bytes @brunoerg Yeah, I am not sure if the name should be something like `no_burning` or something like `reject_nonzero_returns`

MarcoFalke commented on src/rpc/mempool.cpp:
```diff
@@ -42,6 +42,7 @@ static RPCHelpMan sendrawtransaction()
             {"maxfeerate", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(DEFAULT_MAX_RAW_TX_FEE_RATE.GetFeePerK())},
              "Reject transactions whose fee rate is higher than the specified value, expressed in " + CURRENCY_UNIT +
                  "/kvB.\nSet to 0 to accept any fee rate.\n"},
+            {"no_burning", RPCArg::Type::BOOL, RPCArg::Default{true}, "If true, reject transactions with non-zero OP_RETURN outputs, which are unspendable and a 'burning' of coins."},
```
I wonder if this can indicate the maximum value for a data output (default: 0) instead of a true/false

Meru852 replied:
marco wheres your responsiblelity of your own words that you said in our chat by telegram?
you already receive 0.15 btc from me

glozow replied:
I think`max_burn_amount` is the right name here, similar to `maxfeerate`, since that's what this check is most similar to.

davidgumberg replied:
Thanks for the suggestion, I have changed the parameter to `maxburnamount` (I avoided snake case in the json parameter to keep consistent with `maxfeerate` and used snake case for our c++ variable `max_burn_amount`) and ensure that our `nValue` is not greater.

MarcoFalke commented on src/rpc/mempool.cpp:
```diff
@@ -61,12 +62,24 @@ static RPCHelpMan sendrawtransaction()
             RPCTypeCheck(request.params, {
                 UniValue::VSTR,
                 UniValueType(), // VNUM or VSTR, checked inside AmountFromValue()
+                UniValue::VBOOL
```
```suggestion
                UniValue::VBOOL,
```

Ellajoke replied:
Hello

davidgumberg replied:
Thanks, I corrected this.

theStack commented:
Concept ACK

davidgumberg commented:
I have some questions:

1. Should the `max_burn_amount` parameter also be added to the `testmempoolaccept` call? I am not 100% sure about the use case for this call, in particular why it has the `maxfeerate` parameter. Is the `maxfeerate` parameter present because this call is a more general "would `sendrawtransaction` work on this tx" than "will the mempool accept this"?
2. I am still trying to understand the BTC transaction language, so I am unsure about whether or not it is possible for `OP_RETURN` to appear somewhere other than at the top of the `scriptPubKey` in a valid tx. And if it can, then does the presence of `OP_RETURN` *anywhere* in the `scriptPubKey` *guarantee* that the value of that output is unspendable? If both are true, then we would need to check every non-data byte of the `scriptPubKey` for an `OP_RETURN`. If the former is true and the latter is false, something more complicated might have to happen.
3. This may be outside of the scope of this pull request, and based on a foolish misunderstanding, but is there any way that we validate that the default parameter promised in the RPC description is actually the default value?

davidgumberg commented:
Also, I apologize for the obscene number of force pushes when my first edit fails CI because of a silly mistake, I realize this makes review more of a pain, going forward I'll run the test suite locally before force pushing.

1440000bytes commented on src/rpc/mempool.cpp:
```diff
@@ -41,7 +41,10 @@ static RPCHelpMan sendrawtransaction()
+                 "/kvB.\nSet to 0 to accept any fee rate."},
+            {"maxburnamount", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(0)},
+             "Reject transactions with OP_RETURN outputs greater than the specified value, expressed in " + CURRENCY_UNIT + ".\n"
+             "If set to 0, all transactions with non-zero OP_RETURN outputs, which are unspendable and a 'burning' of coins, will be rejected.\n"},
```
```suggestion
             "Default value is 0 which disallows non-zero OP RETURN outputs.\n"},
```

1440000bytes commented on src/util/error.cpp:
```diff
@@ -33,6 +33,8 @@ bilingual_str TransactionErrorString(const TransactionError err)
         case TransactionError::MAX_FEE_EXCEEDED:
             return Untranslated("Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)");
+        case TransactionError::MAX_BURN_EXCEEDED:
+            return Untranslated("OP_RETURN output exceeds maximum set by maxburnamount");
```
```suggestion
            return Untranslated("OP_RETURN output exceeds maximum configured by user (-maxburnamount)");
```

davidgumberg replied:
There's no commandline option like for `maxfeerate` so there shouldn't be a hyphen, but I'll change to configured by user, thanks.

1440000bytes reviewed (commented):
Tested `maxburnamount` parameter added in `sendratransaction` RPC and it works as expected: 

1. Create a transaction in electrum (signet) with `OP_RETURN 74657374` in 'payto' and `0.0001` in amount. Sign and copy the transaction hex.

2. Try to broadcast transaction in bitcoin core with default settings and custom burn amount:

```
$ bitcoin-cli sendrawtransaction 020000000001019bc0177bd2816b10daecd76401ccbacffa8b4d83c3a1c8f6461fb0c88633de840000000000fdffffff021027000000000000066a047465737430f4080000000000160014d77709c50f964d6ccc668c3345a0dd9ba083e05c0247304402206a0d0cc140d9e99caccc776474669a57bac3f7e1c605b5b5ca7815d021e65ccd02200172ef8fe2693e2d4564b0aec5ebc28cc3c75f8475508a31b5a9c2271658b9c001210205e865e6c9b71d16ce7ba97b999db8a08115648abbaeca59bd462b0a8fd0054a209d0100

OP_RETURN output exceeds maximum set by maxburnamount (code -25)

$ bitcoin-cli -named sendrawtransaction hexstring=020000000001019bc0177bd2816b10daecd76401ccbacffa8b4d83c3a1c8f6461fb0c88633de840000000000fdffffff021027000000000000066a047465737430f4080000000000160014d77709c50f964d6ccc668c3345a0dd9ba083e05c0247304402206a0d0cc140d9e99caccc776474669a57bac3f7e1c605b5b5ca7815d021e65ccd02200172ef8fe2693e2d4564b0aec5ebc28cc3c75f8475508a31b5a9c2271658b9c001210205e865e6c9b71d16ce7ba97b999db8a08115648abbaeca59bd462b0a8fd0054a209d0100 maxburnamount=0.0001

76fe1211a1059ff9bcf6e2b330631c75b901e0884669ff47be82ec948d84e68c
```

MarcoFalke commented:
> I am still trying to understand the BTC transaction language, so I am unsure about whether or not it is possible for OP_RETURN to appear somewhere other than at the top of the scriptPubKey in a valid tx. And if it can, then does the presence of OP_RETURN anywhere in the scriptPubKey guarantee that the value of that output is unspendable? If both are true, then we would need to check every non-data byte of the scriptPubKey for an OP_RETURN. If the former is true and the latter is false, something more complicated might have to happen.

I think you can just use the solver to check for NULL_DATA outputs. Everything else with an OP_RETURN is nonstandard and is thus rejected anyway (regardless of amount), unless the user modified the settings or the source code.

davidgumberg commented:
I have changed the `OP_RETURN` check to use `Solver` and I have added the tests, so I'm removing `[WIP]` from the PR.

davidgumberg commented:
As a side note: would it make sense to have an overloaded `Solver` with no Solutions vector parameter for when we only need the `TxoutType`? I count that 5 of the 15 existing calls to Solver don't use the solutions vector data that gets returned:
`AreInputsStandard` in policy.cpp, `CBloomFilter::IsRelevantAndUpdate` in bloom.cpp, `ScriptToUniv` in core_write.cpp, `AvailableCoins` in spend.cpp, and `CWallet::TransactionChangeType` in wallet.cpp

w0xlt reviewed (commented):
The code seems to meet the proposed objective.

But, as mentioned in https://github.com/bitcoin/bitcoin/issues/25899#issuecomment-1229361805,  it's also unclear to me why any application needs to use non-zero op_returns, considering the money used is lost.

It might be good to include in the PR and commit description some real cases where this feature is used.

1440000bytes commented:
> The code seems to meet the proposed objective.
> 
> But, as mentioned in [#25899 (comment)](https://github.com/bitcoin/bitcoin/issues/25899#issuecomment-1229361805), it's also unclear to me why any application needs to use non-zero op_returns, considering the money used is lost.
> 
> It might be good to include in the PR and commit description some real cases where this feature is used.

We as a group of (less than 100 active developers) cannot predict how bitcoin core is used. There can be be use cases where it seems appropriate to burn a small amount.

Its good that this is optional and users can still decide the burn amount with a parameter.

davidgumberg commented:
+1 @1440000bytes 

Given that burned utxo's don't negatively impact the network since `OP_RETURN` tx's are prunable, I think  we should do whatever is possible to  prevent users from mistakenly burning coins, but we should not try to prevent users who are dead-set on burning coins and understand the consequences from doing so. 

As mentioned in #25899, users cannot even craft raw tx's with nonzero `OP_RETURN` outputs using the reference client. The theoretical transactions we are rejecting would have to have been either written in hex by hand or crafted with another client (as @1440000bytes did above) and broadcast by the reference client.

maflcko reviewed (approved):
ACK, lgtm

glozow commented:
Thanks for working on this, I think it is a good improvement to help prevent accidental money loss.
Might want a [release note](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#release-notes)?

1440000bytes reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/25943/commits/83882e65b12ebd4e0a72967dbd8e54520eeecaa5

luke-jr commented on src/rpc/mempool.cpp:
```diff
@@ -41,7 +42,10 @@ static RPCHelpMan sendrawtransaction()
              "Reject transactions whose fee rate is higher than the specified value, expressed in " + CURRENCY_UNIT +
-                 "/kvB.\nSet to 0 to accept any fee rate.\n"},
+                 "/kvB.\nSet to 0 to accept any fee rate."},
+            {"maxburnamount", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(0)},
```
Probably should be an options object instead (can replace maxfeerate).

davidgumberg replied:
Conceptually I agree. I don't have much practical knowledge of how the RPC interface is used in the wild but wouldn't this break existing code that sets a `maxfeerate`? I also see that there are a few other calls that pass multiple options as separate arguments like `convertopsbt` and `gettxoutsetinfo` so maybe there is a case for making a broader change outside of the scope of this PR. 

But, if in practice this would not meaningfully break existing use cases then I will change it to an options object.

davidgumberg replied:
Marking as resolved unless there is disagreement.

luke-jr commented on src/rpc/mempool.cpp:
```diff
@@ -41,7 +42,10 @@ static RPCHelpMan sendrawtransaction()
-                 "/kvB.\nSet to 0 to accept any fee rate.\n"},
+                 "/kvB.\nSet to 0 to accept any fee rate."},
+            {"maxburnamount", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(0)},
+             "Reject transactions with OP_RETURN outputs greater than the specified value, expressed in " + CURRENCY_UNIT + ".\n"
```
"datacarrier" is typically the term used for this rather than "OP_RETURN" (which is an opcode)

davidgumberg replied:
I corrected this, thanks

rajarshimaitra reviewed (commented):
Concept + tACK 83882e65b12ebd4e0a72967dbd8e54520eeecaa5

Just a minor nit..

davidgumberg commented on src/util/error.cpp:
```diff
@@ -33,6 +33,8 @@ bilingual_str TransactionErrorString(const TransactionError err)
         case TransactionError::MAX_FEE_EXCEEDED:
             return Untranslated("Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)");
+        case TransactionError::MAX_BURN_EXCEEDED:
+            return Untranslated("OP_RETURN output exceeds maximum configured by user (maxburnamount)");
```
Thanks for the review, but I disagree with the nit,

``Maximum - n., The highest possible magnitude or quantity of something which is attained, attainable, or customary; an upper limit of magnitude or quantity``

OED Online

rajarshimaitra replied:
It is clear what "maximum" means, but isn't clear what "quantity" it is talking about..  But not a big deal..

davidgumberg committed:
Add test for unspendable transactions and parameter 'maxburnamount' to sendrawtransaction.

'maxburnamount' sets a maximum value for outputs heuristically deemed unspendable including datacarrier scripts that begin with `OP_RETURN`.

davidgumberg commented:
Rebased to master and changed text in rpc description from `OP_RETURN outputs` to `datacarrier outputs` as suggested by @luke-jr .

MarcoFalke commented:
I think you want to replace all occurrences of OP_RETURN output in your patch with datacarrier output, if you replace one?

achow101 commented:
I think this should be extended to also include any provably unspendable outputs. Currently that would be anything with an OP_RETURN, and anything larger than the maximum script size. Using the `Solver` and checking for `NULL_DATA` only finds those things which match the standardness rules for datacarrier, not all OP_RETURN scripts nor anything larger than the max script size. However we have `CScript::IsUnspendable` which does check for those things.

The help text would need to be updated to say "unspendable outputs" rather than "datacarrier" or "OP_RETURN", although it could mention those as examples of unspendable outputs.

MarcoFalke commented:
Nonstandard (spendable and unspendable) transactions are already rejected, so I am not sure how much difference this will make in practice, but no objection to adding more checks either here or in a follow-up. Also, standard transaction outputs might be unprovably unspendable, so if this is changed, it should be clear that the unspendable check is a heuristic at best.

MarcoFalke commented:
It may also check for `HasValidOps`, but I think all of this can be done in a follow-up.

davidgumberg commented:
Rebased, I have adjusted the language to indicate that the check is for provable unspendability, instead of just datacarrier txn's and changed the check to use both `IsUnspendable` as suggested by @achow101 and `HasValidOps` as suggested by @MarcoFalke. I have added new functional tests that check that transactions containing invalid opcodes, or exceeding maximum script size are rejected. Since this change exposes a new RPC option, I have added a release note as suggested by @glozow.

It seems to me that care should be taken that there are never any previously spendable and legitimate transactions that `sendrawtransaction` refuses to send as a result of this check. If needed, what should be done to better ensure this change will not cause any transactions that are expected to be spendable to be rejected?

achow101 commented:
ACK ab8c380a7370e29dfedce1399ad1b2a2d234915e

> It seems to me that care should be taken that there are never any previously spendable and legitimate transactions that `sendrawtransaction` refuses to send as a result of this check. If needed, what should be done to better ensure this change will not cause any transactions that are expected to be spendable to be rejected?

Even if there are, a suitable workaround is to just set `maxburnamount` to a high value. However I believe the check is specific enough that we won't have any issues other than with people who are explicitly trying to burn money. `IsUnspendable` is used by mempool policy and validation, and `HasValidOps` is used by the transaction hex parser (so actually it might be redundant here). These are things that, if they were not working correctly, would cause issues in other parts of the codebase.

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -44,7 +45,11 @@ static RPCHelpMan sendrawtransaction()
+                 "/kvB.\nSet to 0 to accept any fee rate."},
+            {"maxburnamount", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(0)},
+             "Reject transactions with provably unspendable outputs (e.g. 'datacarrier' outputs that use the OP_RETURN opcode) greater than the specified value, expressed in " + CURRENCY_UNIT + ".\n"
+             "If set to 0, all transactions with provably unspendable outputs will be rejected.\n"
```
This isn't true? If set to 0, transactions with provably unspendable outputs are fine as long as the output values are 0. Perhaps remove this line, as we wouldn't want a user to mistakenly think they're filtering out all datacarrier transactions by setting this (also, the same thing happens if this option is left blank since 0 is default).

glozow replied:
Perhaps add "If burning funds through unspendable outputs is desired, increase this value." If a downstream app is (hypothetically) regularly burning money and upgrades their bitcoind, it should be easy to understand what to do to make their error go away.

glozow reviewed (commented):
ACK ab8c380a73 but I think one of the docs is a bit misleading

davidgumberg committed:
Add release note for PR#25943

Co-authored-by: glozow <gloriajzhao@gmail.com>

davidgumberg commented:
Force-pushed to address @glozow review.

glozow commented:
reACK 7013da07fbcddb04abae9759767a9419ab90444c

achow101 commented:
re-ACK 7013da07fbcddb04abae9759767a9419ab90444c
